modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,653 @@
+            return Err(Error::invalid_trace_length(v.len()));
+        }
+
+        let mut trace = vec![];
+        let id_pairs = v.chunks_exact(2).map(|paths| (paths[0], paths[1]));
+        for (pos, (port_id, channel_id)) in id_pairs.rev().enumerate() {
+            let port_id =
+                PortId::from_str(port_id).map_err(|e| Error::invalid_trace_port_id(pos, e))?;
+            let channel_id = ChannelId::from_str(channel_id)
+                .map_err(|e| Error::invalid_trace_channel_id(pos, e))?;
+            trace.push(TracePrefix {
+                port_id,
+                channel_id,
+            });
+        }
+
+        Ok(trace.into())
+    }
+}
+
+impl FromStr for TracePath {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = {
+            let parts: Vec<&str> = s.split('/').collect();
+            if parts.len() == 1 && parts[0].trim().is_empty() {
+                vec![]
+            } else {
+                parts
+            }
+        };
+        parts.try_into()
+    }
+}
+
+impl fmt::Display for TracePath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let path = self
+            .0
+            .iter()
+            .rev()
+            .map(|prefix| prefix.to_string())
+            .collect::<Vec<String>>()
+            .join("/");
+        write!(f, "{}", path)
+    }
+}
+
+/// Indicates whether the sender chain is acting as a source or sink. Each send to any chain other
+/// than the one it was previously received from is a movement forwards in the token's timeline. In
+/// these instances the sender chain is acting as the source zone.
+pub enum Source {
+    Sender,
+    Receiver,
+}
+
+/// A type that contains the base denomination for ICS20 and the source tracing information path.
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct DenomTrace {
+    /// A series of `{port-id}/{channel-id}`s for tracing the source of the token.
+    #[serde(with = "serde_string")]
+    trace_path: TracePath,
+    /// Base denomination of the relayed fungible token.
+    base_denom: Denom,
+}
+
+impl DenomTrace {
+    /// Returns a coin denomination for an ICS20 fungible token in the format
+    /// 'ibc/{Hash(trace_path/base_denom)}'.
+    pub fn hashed(&self) -> HashedDenom {
+        HashedDenom::from(self)
+    }
+
+    /// Removes the specified prefix from the trace path if there is a match, otherwise does nothing.
+    pub fn remove_trace_prefix(&mut self, prefix: &TracePrefix) {
+        self.trace_path.remove_prefix(prefix)
+    }
+
+    /// Adds the specified prefix to the trace path.
+    pub fn add_trace_prefix(&mut self, prefix: TracePrefix) {
+        self.trace_path.add_prefix(prefix)
+    }
+
+    /// Returns `Source::Receiver` if the denomination originally came from the receiving chain and
+    /// `Source::Sender` otherwise.
+    pub fn source_chain(&self, prefix: &TracePrefix) -> Source {
```

Comment:
This method relies on the caller to build the prefix in the appropriate way for its semantics to be correct. In other words, the caller must construct `prefix` from a packet's `source_port` and `source_channel` for the return value of this function to be correct. This is an indication that this function doesn't belong in the `DenomTrace` type. Also, intuitively, a denom trace has no notion of a "source chain", let alone a chain.

We should create an `Ics20Packet` struct, which contains the deserialized `PacketData` as a field. We can then put `source_chain(&self)` as a method of `Ics20Packet`. IMO this is cleaner, as an ICS20 packet has all the information necessary to make that decision.

Bonus: with this proposed change we get the benefit that `process_recv_packet()` doesn't take a `Packet` and `PacketData`, where `PacketData` is contained within `Packet` in an serialized form. It can now take an `Ics20Packet`. That was bothering me :).

Reply:
> We should create an Ics20Packet struct, which contains the deserialized PacketData as a field. 

This method is also called with the port and channel fields of `MsgTransfer` in `send_transfer()`. 
IMHO, we would be better off avoiding avoiding the complexity and following ibc-go's API - free-standing functions that return `bool`s as simple as -> 
```rust
fn is_sender_chain_source(source_port: &PortId, source_channel: ChannelId, denom: &PrefixedDenom) -> bool {
    !is_receiver_chain_source(source_port, source_channel, denom)
}

fn is_receiver_chain_source(source_port: &PortId, source_channel: ChannelId, denom: &PrefixedDenom) -> bool {
    // ...
}
```